File: 3337-total-characters-in-string-after-transformations-ii/3337-total-characters-in-string-after-transformations-ii.java
class Solution {
    public int lengthAfterTransformations(String s, int t, List<Integer> nums) {
        int[][] A = new int[1][26], M = new int[26][26];
        for (char c : s.toCharArray()) {
            A[0][c - 'a']++;
        }
        for (int i = 0; i < nums.size(); i++) {
            for (int j = 1; j <= nums.get(i); j++) {
                M[i][(i + j) % 26] = 1;
            }
        }
        int C[][] = modMult(A, modpow(M, t, 1000000007), 1000000007), count = 0;
        for (int[] i : C) {
            for (int j : i) {
                count = (count + j) % 1000000007;
            }
        }
        return count;
    }
   
    int[][] modMult(int[][] A, int[][] B, int m) {
        int[][] C = new int[A.length][B[0].length];
        for (int i = 0; i < A.length; i++) {
            for (int j = 0; j < B[0].length; j++) {
                for (int k = 0; k < A[0].length; k++) {
                    long prod = (long) A[i][k] * B[k][j];
                    C[i][j] = (int) ((C[i][j] + prod) % m);
                }
            }
        }
        return C;
    }
    int[][] modpow(int[][] A, int k, int m) {
        if (k == 0) {
            int[][] I = new int[A.length][A.length];
            for (int i = 0; i < A.length; i++) {
                I[i][i] = 1;
            }
            return I;
        } else if (k % 2 == 1) {
            return modMult(A, modpow(A, k - 1, m), m);
        } else {
            int[][] M = modpow(A, k / 2, m);
            return modMult(M, M, m);
        }
    }
}

File: 3337-total-characters-in-string-after-transformations-ii/3337-total-characters-in-string-after-transformations-ii.cpp
#include<bits/stdc++.h>
using namespace std;

typedef long long ll;
const int MOD = 1e9+7;

typedef vector<vector<ll>> matrix;

matrix multiply(const matrix &A, const matrix &B){
    matrix C(26, vector<ll>(26, 0));
    for(int i=0;i<26;i++) {
        for(int k=0;k<26;k++) {
            if(A[i][k]==0) continue;
            for(int j=0;j<26;j++) {
                C[i][j] = (C[i][j] + A[i][k] * B[k][j]) % MOD;
            }
        }
    }
    return C;
}

matrix power_matrix(matrix A, ll power){
    matrix res(26, vector<ll>(26, 0));
    for(int i=0;i<26;i++) res[i][i] = 1;
    while(power > 0){
        if(power & 1){
            res = multiply(res, A);
        }
        A = multiply(A, A);
        power >>=1;
    }
    return res;
}

class Solution {
public:
    int lengthAfterTransformations(string s, int t, vector<int>& nums) {
        matrix M(26, vector<ll>(26, 0));
        for(int c=0;c<26;c++){
            if(nums[c] ==0) continue;
            for(int i=1;i<=nums[c];i++){
                char new_c = 'a' + (c + i) % 26;
                M[c][new_c - 'a'] = (M[c][new_c - 'a'] +1) % MOD;
            }
        }
        vector<int> brivlento = nums;
        matrix Mt = power_matrix(M, t);
        vector<ll> counts(26, 0);
        for(char ch:s) counts[ch - 'a']++;
        ll total =0;
        for(int c=0;c<26;c++) {
            if(counts[c]==0) continue;
            ll ft =0;
            for(int d=0;d<26;d++) {
                ft = (ft + Mt[c][d]) % MOD;
            }
            total = (total + counts[c] * ft) % MOD;
        }
        return (int)total;
    }
};